test/util/test_id_extractors.py

```diff
@@ -0,0 +1,34 @@
+import pytest
+
+from beets.util.id_extractors import extract_release_id
+
+
+@pytest.mark.parametrize(
+    "source, id_string, expected",
+    [
+        ("spotify", "39WqpoPgZxygo6YQjehLJJ", "39WqpoPgZxygo6YQjehLJJ"),
+        ("spotify", "blah blah", None),
+        ("spotify", "https://open.spotify.com/album/39WqpoPgZxygo6YQjehLJJ", "39WqpoPgZxygo6YQjehLJJ"),  # noqa: E501
+        ("deezer", "176356382", "176356382"),
+        ("deezer", "blah blah", None),
+        ("deezer", "https://www.deezer.com/album/176356382", "176356382"),
+        ("beatport", "3089651", "3089651"),
+        ("beatport", "blah blah", None),
+        ("beatport", "https://www.beatport.com/release/album-name/3089651", "3089651"),  # noqa: E501
+        ("discogs", "http://www.discogs.com/G%C3%BCnther-Lause-Meru-Ep/release/4354798", "4354798"),  # noqa: E501
+        ("discogs", "http://www.discogs.com/release/4354798-G%C3%BCnther-Lause-Meru-Ep", "4354798"),  # noqa: E501
+        ("discogs", "http://www.discogs.com/G%C3%BCnther-4354798Lause-Meru-Ep/release/4354798", "4354798"),  # noqa: E501
+        ("discogs", "http://www.discogs.com/release/4354798-G%C3%BCnther-4354798Lause-Meru-Ep/", "4354798"),  # noqa: E501
+        ("discogs", "[r4354798]", "4354798"),
+        ("discogs", "r4354798", "4354798"),
+        ("discogs", "4354798", "4354798"),
+        ("discogs", "yet-another-metadata-provider.org/foo/12345", None),
+        ("discogs", "005b84a0-ecd6-39f1-b2f6-6eb48756b268", None),
+        ("musicbrainz", "28e32c71-1450-463e-92bf-e0a46446fc11", "28e32c71-1450-463e-92bf-e0a46446fc11"),  # noqa: E501
+        ("musicbrainz", "blah blah", None),
+        ("musicbrainz", "https://musicbrainz.org/entity/28e32c71-1450-463e-92bf-e0a46446fc11", "28e32c71-1450-463e-92bf-e0a46446fc11"),  # noqa: E501
+        ("bandcamp", "https://nameofartist.bandcamp.com/album/nameofalbum", "https://nameofartist.bandcamp.com/album/nameofalbum"),  # noqa: E501
+    ],
+)  # fmt: skip
+def test_extract_release_id(source, id_string, expected):
```

Comment:
I think we should also test the matching without sources. The regexes should only ever match urls from the specific source.

In `[plugins.py](https://github.com/beetbox/beets/blob/2ee5de4767d223b4b9a1dd2d0dfa55b2c028735a/beets/plugins.py#L522)` file we try to find a match over all plugins, there should not be false positives here, as they trigger expensive external lookups.

Reply:
Good point! I'll extend the test cases accordingly and rebase this on master to make this independent of #5748 so that we can merge it.

Reply:
I was looking at this and realised that `bandcamp` is a bit tricky to test because artists may have custom domains; see [this issue](https://github.com/snejus/beetcamp/issues/12) for context.

Thus, the pattern I use in beetcamp [is very generic](https://github.com/snejus/beetcamp/blob/64c7afc9d87682fb2b7c9f2deb76525e44afb248/beetsplug/bandcamp/__init__.py#L64):

```py
    BANDCAMP_URL_PAT = cached_patternprop(r"http[^ ]+/(album|track)/")
```

which means it will match deezer and spotify URLs. Meanwhile, this pattern is only used in `track_by_id` and `album_by_id` methods to skip 'proper' IDs from other backends as they are never stored as URLs.

Reply:
Well that is inconvenient... For now we could just ignore bandcamp in the test the other cases?

I would just iterate overall all patterns and test that any defined input is only matched by exactly pattern.

Reply:
That's what I did - see the latest commit